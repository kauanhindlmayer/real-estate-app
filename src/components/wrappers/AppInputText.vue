<script lang="ts" setup>
import InputText from 'primevue/inputtext'
import { ref } from 'vue'
import { useUid } from '@/composables/useUid'

interface IProps {
  label?: string
  required?: boolean
}

const props = withDefaults(defineProps<IProps>(), {
  label: '',
  required: false
})

const modelValue = defineModel<String | Number>()
const errorMessage = ref('')
const uid = useUid()

function isValid(): boolean {
  if (props.required && !modelValue.value) {
    errorMessage.value = 'This field is required'
    return false
  }
  errorMessage.value = ''
  return true
}

defineExpose({
  isValid
})
</script>

<template>
  <label v-if="label" :for="uid">
    {{ label }}
    <span v-if="required" class="p-error"> *</span>
  </label>
  <InputText
    v-bind="$attrs"
    v-model="modelValue"
    :id="uid"
    :aria-labelledby="errorMessage ? `${uid}-error` : undefined"
    :aria-invalid="!!errorMessage"
  />
  <small class="p-error" v-if="errorMessage" :id="`${uid}-error`" aria-live="assertive">
    {{ errorMessage }}
  </small>
</template>
