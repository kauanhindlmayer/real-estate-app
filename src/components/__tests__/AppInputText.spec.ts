import { describe, it, expect, beforeEach } from 'vitest'
import { VueWrapper, mount } from '@vue/test-utils'

import PrimeVue from 'primevue/config'
import AppInputText from '@/components/wrappers/AppInputText.vue'

describe('AppInputText', () => {
  let wrapper: VueWrapper<any>

  beforeEach(() => {
    wrapper = mount(AppInputText, {
      global: {
        plugins: [PrimeVue]
      }
    })
  })

  it('renders properly', async () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders the label when provided', async () => {
    wrapper.setProps({ label: 'Name' })
    await wrapper.vm.$nextTick()
    const labelElement = wrapper.find('label')
    expect(labelElement.exists()).toBe(true)
    expect(labelElement.text()).toBe('Name')
  })

  it('does not render the label when not provided', async () => {
    const labelElement = wrapper.find('label')
    expect(labelElement.exists()).toBe(false)
  })

  it('renders the error message when input is invalid', async () => {
    wrapper.setProps({ modelValue: '', label: 'Name', required: true })
    await wrapper.vm.$nextTick()
    wrapper.vm.isValid()
    await wrapper.vm.$nextTick()
    const errorMessageElement = wrapper.find('small.p-error')
    expect(errorMessageElement.exists()).toBe(true)
    expect(errorMessageElement.text()).toBe('This field is required')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('does not render the error message when input is valid', async () => {
    wrapper.setProps({ modelValue: 'John Doe', required: true })
    await wrapper.vm.$nextTick()
    wrapper.vm.isValid()
    await wrapper.vm.$nextTick()
    const errorMessageElement = wrapper.find('small.p-error')
    expect(errorMessageElement.exists()).toBe(false)
  })
})
