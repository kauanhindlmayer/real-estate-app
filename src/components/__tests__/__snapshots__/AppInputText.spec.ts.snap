// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppInputText > renders properly 1`] = `
"<!--v-if-->
<input class="p-inputtext p-component" aria-invalid="false" id="0" data-pc-name="inputtext" data-pc-section="root">
<!--v-if-->"
`;

exports[`AppInputText > renders the error message when input is invalid 1`] = `
"<label for="3">Name <span class="p-error"> *</span></label>
<input class="p-inputtext p-component" aria-invalid="true" id="3" data-pc-name="inputtext" data-pc-section="root" aria-labelledby="3-error">
<small class="p-error" id="3-error" aria-live="assertive">This field is required</small>"
`;
